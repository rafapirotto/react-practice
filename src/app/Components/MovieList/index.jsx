import React, { useEffect, useState } from 'react';
import { useDispatch, connect } from 'react-redux';
import PropTypes from 'prop-types';

import { getMovies } from './duck/operations';
import { ERROR, LOADING, SUCCESS } from '../../../common/constants';
import spinner from './assets/spinner.gif';
import error from './assets/error.png';
import Movie from '../Movie';

const propTypes = {
  movies: PropTypes.array.isRequired,
};

const renderMovies = (movies) => (
  <div className="flex mt-8 ml-24 flex-wrap">
    {movies.map(({ poster_url, title }) => (
      <Movie url={poster_url} title={title} />
    ))}
  </div>
);

const renderTitle = (title) => (
  <div>
    <h1 className="ml-24 text-3xl font-semibold">{title}</h1>
  </div>
);

const renderSpinner = () => (
  <div className="mt-40 flex justify-center">
    <img src={spinner} alt="spinner" />
  </div>
);

const renderError = () => (
  <div className="mt-20 flex justify-center">
    <img src={error} alt="error" />
  </div>
);

const renderComponentConditionally = (condition, callback) => condition && callback();

const MovieList = ({ movieList }) => {
  const dispatch = useDispatch();
  const [movies, setMovies] = useState([]);
  const { state, content } = movieList;

  useEffect(() => {
    dispatch(getMovies());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (state === SUCCESS) {
      setMovies(content);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [movieList]);

  return (
    <main>
      {renderTitle('My List')}
      {renderComponentConditionally(state === SUCCESS, () => renderMovies(movies))}
      {renderComponentConditionally(state === LOADING, () => renderSpinner())}
      {renderComponentConditionally(state === ERROR, () => renderError())}
    </main>
  );
};

const mapStateToProps = (state) => ({
  movieList: state.movieList,
});

MovieList.propTypes = propTypes;

export default connect(mapStateToProps)(MovieList);
