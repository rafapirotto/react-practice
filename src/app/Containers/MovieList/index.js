import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import { getMovies } from './duck/operations';
import { SUCCESS } from '../../../common/constants';
import { MovieList } from '../../Components';

const MovieListContainer = () => {
  const dispatch = useDispatch();
  const [movies, setMovies] = useState([]);
  const { state, content } = useSelector((state) => state.movies);

  useEffect(() => {
    if (!content) {
      dispatch(getMovies());
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (state === SUCCESS) {
      setMovies(content);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [content]);

  return (
    <main>
      <MovieList movies={movies} state={state} />
    </main>
  );
};

export default MovieListContainer;
