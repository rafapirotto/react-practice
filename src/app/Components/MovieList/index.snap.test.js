import React from 'react';
import renderer from 'react-test-renderer';

import MovieList from '.';

it('Matches the snapshot', () => {
  const component = renderer.create(
    <MovieList
      movies={[{ poster_url: 'url', title: 'title', id: 2 }]}
      state="some-state"
      title="some-title"
      watchingProgress
    />,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
