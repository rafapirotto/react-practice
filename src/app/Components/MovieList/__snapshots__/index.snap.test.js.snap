// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches the snapshot 1`] = `
<div>
  <h1
    className="ml-24 text-3xl font-medium"
  >
    some-title
  </h1>
</div>
`;
